calyx-frontend/src/attribute.rs

```diff
@@ -1,146 +1,198 @@
-                )));
-            }
-            attrs.insert(k, v);
-        }
-        Ok(Attributes {
-            attrs,
-            span: GPosIdx::UNKNOWN,
-        })
+impl std::fmt::Display for NumAttr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.as_ref())
     }
 }
 
-impl WithPos for Attributes {
-    fn copy_span(&self) -> GPosIdx {
-        self.span
+#[derive(AsRefStr, Clone, Copy, Hash, PartialEq, Eq, Debug)]
+#[allow(non_camel_case_types)]
+/// Internal attributes that cannot be parsed back from the IL.
+pub enum InternalAttr {
+    DEAD,
+    NODE_ID,
+    BEGIN_ID,
+    END_ID,
+    ST_ID,
+    LOOP,
+    START,
+    END,
+}
+impl From<InternalAttr> for Attribute {
+    fn from(attr: InternalAttr) -> Self {
+        Attribute::Internal(attr)
     }
 }
 
-/// Structs that can return an [`Attributes`] instance.
-pub trait GetAttributes {
-    /// Returns an [`Attributes`] instance
-    fn get_attributes(&self) -> &Attributes;
-
-    /// Returns a mutable [`Attributes`] instance
-    fn get_mut_attributes(&mut self) -> &mut Attributes;
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+/// Defines the known attributes that can be attached to IR nodes.
+/// All caps names represent attributes that are internal to the compiler and
+/// cannot be parsed back.
+pub enum Attribute {
+    Bool(BoolAttr),
+    Num(NumAttr),
+    Internal(InternalAttr),
+    /// Unknown attribute. Should not appear in the Calyx codebase.
+    /// Useful for other frontends using Calyx
+    Unknown(Id),
 }
-
-impl Attributes {
-    /// Add a new attribute
-    pub fn insert<S>(&mut self, key: S, val: u64)
-    where
-        S: Into<Id>,
-    {
-        self.attrs.insert(key.into(), val);
+impl std::fmt::Display for Attribute {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Attribute::Bool(b) => write!(f, "{}", b.as_ref()),
+            Attribute::Num(n) => write!(f, "{}", n.as_ref()),
+            Attribute::Internal(i) => write!(f, "{}", i.as_ref()),
+            Attribute::Unknown(s) => write!(f, "{}", s),
+        }
     }
-
-    /// Get the value associated with an attribute key
-    pub fn get<S>(&self, key: S) -> Option<&u64>
-    where
-        S: Into<Id>,
-    {
-        self.attrs.get(&key.into())
+}
+impl FromStr for Attribute {
+    type Err = Error;
+
+    fn from_str(s: &str) -> CalyxResult<Self> {
+        if let Ok(b) = BoolAttr::from_str(s) {
+            Ok(Attribute::Bool(b))
+        } else if let Ok(n) = NumAttr::from_str(s) {
+            Ok(Attribute::Num(n))
+        } else {
+            // Reject attributes that all caps since those are reserved for internal attributes
+            if s.to_uppercase() == s {
+                return Err(Error::misc(format!("Invalid attribute: {}. All caps attributes are reserved for internal use.", s)));
+            }
+            Ok(Attribute::Unknown(s.into()))
+        }
```

Comment:
Just marking this as kind of the "essence" of the change: we are recognizing a list of attributes by name and turning them into special enum values, with a fallback variant for all others.